grommet/grommet
# Add test coverage enforcement via Jest config

Signed-off-by: brendon.caulkins <brendon.caulkins@gmail.com>

#### What does this PR do?

This small change to `package.json` enables testing coverage enforcement via Jest, as a way to ward off future needs to swarm test coverage, like #4539 .  This approach is just one solution, and has the following benefits:

* Coverage minimums as code, so they are public, and can only be updated via PRs
* Requires no extra packages or integrations
* Works locally, including in pre-commit hooks
* Supports many configuration options beyond what's done in this PR: https://jestjs.io/docs/en/configuration#coveragethreshold-object

#### Where should the reviewer start?

`package.json` and the Jest documentation: https://jestjs.io/docs/en/configuration#coveragethreshold-object

#### What testing has been done on this PR?

I set one of the thresholds below the current amount and the test run failed.  This is shown below in the included screenshot

#### How should this be manually tested?

You could alter the thresholds and see the failure due to being below, or add enough code somewhere to cause the coverage numbers to drop below the configured amount.

#### Any background context you want to provide?

Enforcing code coverage minimums ensures ground isn't lost while development moves forward.  It's largely a safeguard and way to establish a standard, but it's no replacement for active pursuit of comprehensive coverage.

#### What are the relevant issues?

See #4539 

#### Screenshots (if appropriate)

Example of a failure, locally (induced by artificially inflating the minimum).  This will also happen in the CI Pipeline, enforcing these minimums for PRs.
![grommet-jest-config-threshold-example](https://user-images.githubusercontent.com/32872899/95514692-5cae5d80-098a-11eb-883c-a7a715b68fcf.PNG)

#### Do the grommet docs need to be updated?

Not for consumers.  The contributing guidelines may need to be updated, if you wish to call out testing specifically.  testing isn't mentioned in the contributing docs, but the current pre-commit hooks are a solid safeguard.

#### Should this PR be mentioned in the release notes?

No.

#### Is this change backwards compatible or is it a breaking change?

This should be 100% backwards compatible and non-breaking as there is no functional code affected.  

**However**, if there are large PRs sitting around when this goes through, they could be surprised if they don't have adequate coverage.  Depending on repo configuration (whether or not PRs must be up-to-date with master before merging), a large PR that has already passed the pipeline checks and is approved could push master below these thresholds and cause master to break.  It may be worth holding off on this until any such large pieces are in, or lowering the values to an agreed upon minimum before merging.
